src/util/code_observer.h

```diff
@@ -0,0 +1,130 @@
+/*******************************************************************\
+
+Module: Misc Utilities
+
+Author: Kurt Degiorgio
+
+\*******************************************************************/
+
+/// \file
+/// code_observer.h
+
+/// This file contains a couple of utility classes that should be used when one
+/// needs to read all codet's of a particular type from a given symbol
+/// table.  This saves you from having to manually iterate through the symbol
+/// table.  These classes essentially implement the observable pattern.
+///
+/// In general, you first register one or more observers with
+/// const_code_observert. The observer is a callback that is
+/// associated with a specific irep_idt. This is used to determine for which
+/// codet's the observer will be notified. The second step involves calling
+/// visit_symbols which will in turn iterate through every codet inside the
+/// given symbol-table and notify registered observers only if the id of the
+/// codet in question matches the associated id of the observer.
+///
+/// Example: Examining all codet(ID_function_call)
+///
+/// \code
+///
+/// void A::examine_function_calls(const codet &code)
+/// {
+///   PRECONDITION(code.get_statement()==ID_function_call);
+///   ...
+/// }
+///
+/// ...
+///
+/// const_code_observert code_observer;
+/// code_observer.register_observer(ID_function_call,
+///     &A::exmaine_function_calls,
+///     this,
+///     std::placeholders::_1);
+/// code_observer.visit_symbols(symbol_table);
+///
+/// ...
+///
+/// \endcode
+///
+/// Note': each unique irep_idt can only be associated with one observer.
+
```

Comment:
Please add a note why this doesn't support modifying symbol table entries and what the impact of using this observer for modifying symbol table entries is.

Reply:
done